homeassistant/components/numato/binary_sensor.py

```diff
@@ -0,0 +1,107 @@
+"""Binary sensor platform integration for Numato USB GPIO expanders."""
+import logging
+
+from numato_gpio import NumatoGpioError
+
+from homeassistant.components.binary_sensor import BinarySensorDevice
+from homeassistant.const import DEVICE_DEFAULT_NAME
+
+from . import (
+    CONF_BINARY_SENSORS,
+    CONF_DEVICES,
+    CONF_ID,
+    CONF_INVERT_LOGIC,
+    CONF_PORTS,
+    DATA_API,
+    DOMAIN,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Set up the configured Numato USB GPIO binary sensor ports."""
+    if discovery_info is None:
+        return
+    binary_sensors = []
+    devices = hass.data[DOMAIN][CONF_DEVICES]
+    for device in [d for d in devices if CONF_BINARY_SENSORS in d]:
+        device_id = device[CONF_ID]
+        platform = device[CONF_BINARY_SENSORS]
+        invert_logic = platform[CONF_INVERT_LOGIC]
+        ports = platform[CONF_PORTS]
+        for port, port_name in ports.items():
+            binary_sensors.append(
+                NumatoGPIOBinarySensor(
+                    port_name,
+                    device_id,
+                    port,
+                    invert_logic,
+                    hass.data[DOMAIN][DATA_API],
+                )
+            )
+    add_entities(binary_sensors, True)
+
+
+class NumatoGPIOBinarySensor(BinarySensorDevice):
+    """Represents a binary sensor (input) port of a Numato GPIO expander."""
+
+    def __init__(self, name, device_id, port, invert_logic, api):
+        """Initialize the Numato GPIO based binary sensor object."""
+        # pylint: disable=no-member
+        self._name = name or DEVICE_DEFAULT_NAME
+        self._device_id = device_id
+        self._port = port
+        self._invert_logic = invert_logic
+        self._state = None
+        self._api = api
+
+    async def async_added_to_hass(self):
+        """Configure the device port as a binary sensor with edge detection."""
+        try:
+            await self.hass.async_add_executor_job(
+                self._api.setup_input, self._device_id, self._port
+            )
+
+            def read_gpio(port, level):
+                self._state = level
+                self.schedule_update_ha_state()
+
+            await self.hass.async_add_executor_job(
+                self._api.edge_detect, self._device_id, self._port, read_gpio
+            )
+        except NumatoGpioError as err:
```

Comment:
We should move the setup before instantiating the entity so we don't risk adding an entity that doesn't work.

If the GPIO setup fails we log an error and continue before adding the entity.

We can use our dispatch helper to signal the entity update callback from a callback function that we create in `setup_platform`.

```py
NUMATO_SIGNAL = "numato_signal_{}_{}"


def setup_platform(...):
    ...
    def read_gpio(port, level):
        """Send signal to entity to have it update state."""
        dispatcher_send(hass, NUMATO_SIGNAL.format(device_id, port), level)

    api = hass.data[DOMAIN][DATA_API]
    for ...:
        for ...:
            try:
                api.setup_input(device_id, port)
                api.edge_detect(device_id, port, read_gpio)
            except NumatoGpioError as err:
                ...
                continue
            binary_sensors.append(NumatoGPIOBinarySensor(...))

    add_entities(binary_sensors, True)
            

class NumatoGPIOBinarySensor(BinarySensorDevice):
    ...

    async def async_added_to_hass(self):
        """Connect state update callback."""
        self.async_on_remove(
            async_dispatcher_connect(
                self.hass,
                NUMATO_SIGNAL.format(self._device_id, self._port),
                self._async_update_state
            )
        )

	@callback
    def _async_update_state(self, level):
		"""Update entity state."""
        self._state = level
        self.async_write_ha_state()
```

Reply:
@MartinHjelmare: I agree, it's better to configure the device during setup, so if the device config fails (e.g. because it's not connected), the entity is not created. Although I'd like a dynamic behavior better like having an entity which is not available, but can be made available at later point in time - plug and play so to say.

Also in your example you abandon the name from the configuration but you use generic names  "numato_signal_{}_{}" instead. I'd rather stick with the configured names because the
wiring at the device ports isn't  dynamic. You decide which sensor or switch to wire to which port, solder some pullup/down resistors to inputs. Same as creating an ESPHome device. Its naming will remain constant. Because the semantics of the ports/functions is literally hard-wired and I don't want the user to be able to change e.g. a grounded input port to an output because, he'd burn down his device through a short circuit when switching the output on.
The configured name indicates the semantics which a generic name doesn't.

So if you don't mind, I'll change to early setup but continue to use port names from the configuration.

Reply:
I'm not following about the name. The `"numato_signal_{}_{}"` is just a template for the signal.

Reply:
Okay, used the unique name for the dispatcher, but not the entity name. Acceptable?

Reply:
Sorry, I was just a little confused. You didn't actually propose to use that dispatcher signal name as entity name. Dispatcher is new to me, just to excuse myself ;-)